### PR TITLE
parse aggregationProjectionSegment with expressionSegment

### PR DESCRIPTION
**Pull request by cheese8**
For #17987.

aggregate-functions's syntax as below:

[SUM([DISTINCT] expr) [over_clause]]

Ref: https://dev.mysql.com/doc/refman/8.0/en/aggregate-functions.html

While the `**expr**` is in `table.column` format, the parser **did not support** to `parse it into ColumnSegment` for now. This PR is plan to support this sutiation.

Changes proposed in this pull request:
- parse aggregationProjectionSegment with columnSegment 
- support mysql, oracle, sql92, sqlserver

### CODE REVIEW

**Review by strongduanmu — CHANGES_REQUESTED** [review]
Hi @cheese8, can you describe the problem with parse in original issue?

**Comment by cheese8** [comment]
> Hi @cheese8, can you describe the problem with parse in original issue?

Yes, while the **expr** is in table.column format, the parser did not support for now. This PR is plan to support this sutiation.

**Comment by tuichenchuxin** [comment]
I think expr has contains columRef. Maybe we should found why this is not parsed to column.

**Review by tuichenchuxin — CHANGES_REQUESTED** [review]
I think expr has contains columRef. Maybe we should found why this is not parsed to column.

File: shardingsphere-sql-parser/shardingsphere-sql-parser-statement/src/main/java/org/apache/shardingsphere/sql/parser/sql/common/segment/dml/item/AggregationProjectionSegment.java [hunk 1]
```diff
@@ -40,6 +40,9 @@ public class AggregationProjectionSegment implements ProjectionSegment, AliasAva
     
     private final String innerExpression;
     
+    @Setter
+    private ExpressionSegment expression;
```
Comment (tuichenchuxin) [inline shardingsphere-sql-parser/shardingsphere-sql-parser-statement/src/main/java/org/apache/shardingsphere/sql/parser/sql/common/segment/dml/item/AggregationProjectionSegment.java:44]: maybe have multi expression.
```
private final Collection<ExpressionSegment> parameters = new LinkedList<>();
```
This maybe better.

**Review by tuichenchuxin — CHANGES_REQUESTED** [review]

File: shardingsphere-sql-parser/shardingsphere-sql-parser-dialect/shardingsphere-sql-parser-mysql/src/main/java/org/apache/shardingsphere/sql/parser/mysql/visitor/statement/impl/MySQLStatementSQLVisitor.java [hunk 2]
```diff
@@ -773,10 +773,33 @@ public final ASTNode visitAggregationFunction(final AggregationFunctionContext c
...
+    
+    private Collection<ExpressionSegment> getExpressions(final AggregationFunctionContext ctx) {
+        if (null == ctx.expr()) {
+            return Collections.emptyList();
         }
-        return new AggregationDistinctProjectionSegment(ctx.getStart().getStartIndex(), ctx.getStop().getStopIndex(), type, innerExpression, getDistinctExpression(ctx));
+        Collection<ExpressionSegment> result = new LinkedList<>();
+        for (ExprContext each : ctx.expr()) {
+            ASTNode visitNode = visit(each);
+            if (visitNode instanceof ColumnSegment) {
```
Comment (tuichenchuxin) [inline shardingsphere-sql-parser/shardingsphere-sql-parser-dialect/shardingsphere-sql-parser-mysql/src/main/java/org/apache/shardingsphere/sql/parser/mysql/visitor/statement/impl/MySQLStatementSQLVisitor.java:794]: visit exprContext is always expressionSegment, So there is no need to cast to ColumnSegment or others.

**Review by tuichenchuxin — CHANGES_REQUESTED** [review]

File: shardingsphere-sql-parser/shardingsphere-sql-parser-dialect/shardingsphere-sql-parser-mysql/src/main/java/org/apache/shardingsphere/sql/parser/mysql/visitor/statement/impl/MySQLStatementSQLVisitor.java (hunk 2 above)
Comment (tuichenchuxin) [inline shardingsphere-sql-parser/shardingsphere-sql-parser-dialect/shardingsphere-sql-parser-mysql/src/main/java/org/apache/shardingsphere/sql/parser/mysql/visitor/statement/impl/MySQLStatementSQLVisitor.java]: There is no need to judge instance of. Can cast to expressionSegment directly, and add to list.

**Review by tuichenchuxin — CHANGES_REQUESTED** [review]

File: shardingsphere-sql-parser/shardingsphere-sql-parser-dialect/shardingsphere-sql-parser-postgresql/src/main/java/org/apache/shardingsphere/sql/parser/postgresql/visitor/statement/impl/PostgreSQLStatementSQLVisitor.java [hunk 3]
```diff
@@ -504,10 +504,22 @@ public ASTNode visitBExpr(final BExprContext ctx) {
...
+        AggregationProjectionSegment projectionSegment = new AggregationProjectionSegment(ctx.getStart().getStartIndex(), ctx.getStop().getStopIndex(), type, innerExpression);
+        projectionSegment.getParameters().addAll(getExpressions(ctx));
+        return projectionSegment;
+    }
+    
+    private Collection<ExpressionSegment> getExpressions(final FuncApplicationContext ctx) {
+        if (null == ctx.funcArgList()) {
+            return Collections.emptyList();
+        }
+        return Collections.singletonList((ExpressionSegment) visit(ctx.funcArgList()));
```
Comment (tuichenchuxin) [inline shardingsphere-sql-parser/shardingsphere-sql-parser-dialect/shardingsphere-sql-parser-postgresql/src/main/java/org/apache/shardingsphere/sql/parser/postgresql/visitor/statement/impl/PostgreSQLStatementSQLVisitor.java:522]: argList could return multi expressionSegments. We should implement visitArgList method.

File: shardingsphere-test/shardingsphere-parser-test/src/main/resources/case/dml/select-aggregate.xml [hunk 4]
```diff
@@ -26,6 +26,15 @@
         </projections>
     </select>
 
+    <select sql-case-id="select_sum_column">
+        <from>
+            <simple-table name="t_order" start-index="48" stop-index="54" />
+        </from>
+        <projections start-index="7" stop-index="41">
+            <aggregation-projection type="SUM" inner-expression="(t_order.user_id)" alias="user_id_sum" start-index="7" stop-index="26" />
```
Comment (tuichenchuxin) [inline shardingsphere-test/shardingsphere-parser-test/src/main/resources/case/dml/select-aggregate.xml:34]: Maybe we need to extract params and judge.

**Review by tuichenchuxin — CHANGES_REQUESTED** [review]

File: shardingsphere-test/shardingsphere-parser-test/src/main/resources/case/dml/select-aggregate.xml [hunk 5]
```diff
@@ -26,6 +26,21 @@
...
 
+    <select sql-case-id="select_sum_column">
+        <from>
+            <simple-table name="t_order" start-index="48" stop-index="54" />
+        </from>
+        <projections start-index="7" stop-index="41">
+            <aggregation-projection type="SUM" inner-expression="(t_order.user_id)" alias="user_id_sum" start-index="7" stop-index="26">
+                <paramters>
+                    <column name="user_id" start-index="17" stop-index="25">
+                        <owner name="t_order" start-index="8" stop-index="16"/>
```
Comment (tuichenchuxin) [inline shardingsphere-test/shardingsphere-parser-test/src/main/resources/case/dml/select-aggregate.xml:37]: This test need to add parameter in ExpectedAggregationProjection.java

**Review by tuichenchuxin — APPROVED** [review]

**Review by strongduanmu — APPROVED** [review]
@cheese8 Good job, merge.

**Pull request merged**